src/meltano/core/plugin/singer/tap.py

```diff
@@ -615,3 +619,53 @@ def catalog_cache_key(self, plugin_invoker):
         key_json = json.dumps(key_dict)
 
         return sha1(key_json.encode()).hexdigest()  # noqa: S303 S324
+
+    def warn_property_not_found(  # noqa: C901
+        self,
+        rules: list[MetadataRule],
+        catalog: CatalogDict,
+    ):
+        """Validate MetadataRules conforms to discovered Catalog.
+
+        Validate MetadataRules against the tap's discovered Catalog & emit
+        warnings for each rule that does not match any property in catalog.
+
+        The method is intentionally written in a defensive manner so as not to
+        raise exceptions. e.g.
+            * use dict.get rather than []-accessor
+            * filter list comprehensions to remove some elements that should
+              not be used for validation
+
+        Args:
+            rules: List of `MetadataRule`
+            catalog: Discovered Source Catalog
+        """
+        property_rules = [r for r in rules if "*" not in r.tap_stream_id]
+        stream_dict = {
+            stream.get("tap_stream_id"): stream
+            for stream in catalog.get("streams", [])
+            if isinstance(stream, dict)
+        }
+
+        def is_not_star(x):
+            return "*" not in x
+
+        def dict_get(dictionary, key):
+            return dictionary.get(key, {})
+
+        for rule in property_rules:
+            if not (s := stream_dict.get(rule.tap_stream_id)):
```

Comment:
I think this is running into another edge case. `CatalogRule.tap_stream_id` is a list of strings for the [select filter](https://docs.meltano.com/concepts/plugins/#select_filter-extra):

https://github.com/meltano/meltano/blob/6e684f233961be1f1b37550fa0f855e3bc097244/src/meltano/core/plugin/singer/catalog.py#L225.

Reply:
ah, interesting. So `tap_stream_id` can also be a list. I feel like those cases ought to not raise warnings, similar to how warnings are not raised when a wildcard is encountered.

Reply:
we could apply a filter to keep only the rules where `tap_stream_id` is an instance of `str`.